R/multi_fitdist.R

```diff
@@ -1,133 +1,141 @@
+#' the output and calculates the AIC and BIC and ranks the output by model fit
+#' which is determined by the `rank_by` argument.
+#'
+#' @details The vector of models given in the `models` argument needs to be
+#' named with the name of the model argument in the supplied function when it
+#' is not the second argument in that function. All elements of the vector need
+#' to be named, see example.
+#'
+#' The data is assumed to always be the first argument of the function supplied
+#' in `func`, `multi_fitdist()` will not work correct if this is not the case.
+#'
+#' @param data A vector or data frame containing data that is required by the
+#' function specified in `func` argument.
 #' @param models A character string or vector of character strings
-#' specifying the names of the candidate models. This follows the R naming
-#' convention for distributions, the density function is `d[name]`.
+#' specifying the names of the candidate models. The naming of the models
+#' should match those required by the function specified in the `func` argument.
+#' The vector of models should be named with the name of the model argument
+#' from the input function when the argument is not second. See details.
+#' @param func A function (`closure`) used to fit the models.
+#' @param rank_by A character string, either "loglik", "aic" or "bic" to rank
+#' the order of the output data frame. Default is "aic".
 #'
 #' @return A data frame of all models
 #' @export
 #'
 #' @examples
+#' \dontrun{
+#' multi_fitdist(
+#'   data = rgamma(n = 100, shape = 1, scale = 1),
+#'   models = c("gamma", "weibull", "lnorm"),
+#'   func = fitdistrplus::fitdist
+#' )
+#'
+#' # example where the models are named for when the model is not the second
+#' # argument of the function input
 #' multi_fitdist(
 #'   data = rgamma(n = 100, shape = 1, scale = 1),
-#'   models = c("gamma", "weibull", "lnorm")
+#'   models = c(distr = "gamma", distr = "weibull", distr = "lnorm"),
+#'   func = fitdistrplus::fitdist
 #' )
-multi_fitdist <- function(data, models) {
-
-  # fit distributions to data
-  if (is.data.frame(data)) {
-    if (ncol(data) == 2) {
-      res <- lapply(
-        models,
-        fitdistrplus::fitdistcens,
-        censdata = data
+#' }
+multi_fitdist <- function(data,
+                          models,
+                          func,
+                          rank_by = c("aic", "bic", "loglik")) {
+
+  # check input
+  func <- match.fun(func)
+  rank_by <- match.arg(rank_by)
+
+  # try and call function input with args
+  tryCatch(
+    {
+      res <- vector("list", length = length(models))
+      for (i in seq_along(models)) {
+        args <- list(data, models[i])
+        if (!is.null(names(models[i]))) {
+          arg_names <- lapply(args, names)
+          names(args)[2] <- arg_names[2]
+          names(args)[1] <- ""
+        }
+        res[[i]] <- do.call(func, args = args)
+      }
+    },
```

Comment:
```suggestion
      res <- vector("list", length = length(models))
    model_has_names <- all(!is.null(names(models)))
      if (model_has_names) {
        model_label <- unique(names(models))
      for (i in seq_along(models)) {
          args <- list(data, models[[i]])
          names(args)[2] <- model_label
          names(args)[1] <- ""
          res[[i]] <- do.call(func, args = args)
          }
      }else{
        for (i in seq_along(models)) {
          args <- list(data, models[[i]])
          res[[i]] <- do.call(func, args = args)
        }
      }
    },
```

Reply:
You're right that it increases cyclomatic complexity from 5 to 7. One of the key changes I made was to prevent the model names from being checked as many times as the `for` loop in the original code. The proposed change checks it only once with the downside of introducing two for loops. It's probably better to keep the original.

Reply:
Thanks for the suggestion. I'm going to keep the original implementation for now, but it was good to consider alternatives.